debezium-core/src/main/java/io/debezium/connector/base/ChangeEventQueue.java

```diff
@@ -241,15 +244,19 @@ protected void doEnqueue(T record) throws InterruptedException {
         try {
             this.lock.lock();
 
-            while (queue.size() >= maxQueueSize || (maxQueueSizeInBytes > 0 && currentQueueSizeInBytes >= maxQueueSizeInBytes)) {
+            while (running && (queue.size() >= maxQueueSize || (maxQueueSizeInBytes > 0 && currentQueueSizeInBytes >= maxQueueSizeInBytes))) {
                 // signal poll() to drain queue
                 this.isFull.signalAll();
                 // queue size or queue sizeInBytes threshold reached, so wait a bit
                 this.isNotFull.await(pollInterval.toMillis(), TimeUnit.MILLISECONDS);
             }
 
-            queue.enqueue(record);
-            // If we pass a positiveLong max.queue.size.in.bytes to enable handling queue size in bytes feature
+            // If not running, throw InterruptedException to allow graceful exit
+            if (!running) {
```

Comment:
Is `volatile` enough here? What if `running` flag is change between this check and the next `enqueue` call?